pkg/registry/common/begin/nse_server.go

```diff
@@ -92,8 +99,37 @@ func (b *beginNSEServer) Unregister(ctx context.Context, in *registry.NetworkSer
 	}
 	eventFactoryServer, ok := b.Load(id)
 	if !ok {
-		// If we don't have a connection to Close, just let it be
-		return &emptypb.Empty{}, nil
+		var q *queue
+		for {
+			var eventCount int
+			var loaded bool
+			q, loaded = b.queueMap.LoadOrStore(id, &queue{eventCount: 1})
+			if loaded {
+				q.Lock()
+				eventCount = q.eventCount
+
+				if eventCount <= 0 {
+					q.Unlock()
+					continue
+				}
+
+				q.eventCount++
+				q.Unlock()
+			}
+
+			if eventCount > 0 || !loaded {
+				break
+			}
+		}
+
+		q.Lock()
+		_, err := next.NetworkServiceEndpointRegistryServer(ctx).Unregister(ctx, in)
+		q.eventCount--
+		if q.eventCount == 0 {
+			b.queueMap.Delete(id)
+		}
+		q.Unlock()
+		return &emptypb.Empty{}, err
```

Comment:
Is there a reason we hold the lock for the whole `next.Unregister` call?
I would think that we only need it for `q.eventCount` manipulations and `b.queueMap.Delete`.

Reply:
Yes, there is a reason for this. Next chain elements can manipulate some data and if we have several concurrent Unregister events it can lead to data races. We need mutex for the whole Unregister call to let only one event go further.

Reply:
Oh, I have just noticed that you don't use event factory here.
But now I have a different question: what will happen if we get `Register` call with the same ID as this Unregister call?